# Donovan Feist; u0924825; CH EN 2450

# Homework 5: Interpolation

## Problem 1

Consider the data in the class notes that showed the viscosity of air as a function of temperature, μ(T), at atmospheric pressure.

### Part 1

Using linear interpolation, estimate μ at 835 K (by hand).

To interpolate the value of $\mu$ at 835 K, the slope of the line between the two points on either side of 835 K must be determined.

In [1]:
slope=(3.843*10**-5-3.698*10**-5)/(850-800)
print("The slope between the points at 800 K and 850 K has a slope of: ", slope,".")

The slope between the points at 800 K and 850 K has a slope of:  2.9000000000000028e-08 .


This must then be plugged into the equation: $y=slope \times (x-x_1) + y_1$, where $x$ is the point being evaluated (835 K), $x_1$ is the the point on the left of the point in question (800 K), and $y_1$ is the value that corresponds to $x_1$ ($3.698 \times 10^{-5}$).

In [2]:
y=slope*(835-800)+3.698*10**-5
print ("The estimated value of μ at 835 K, using linear interpolation, is: ", y, "N.s/m^2.")

The estimated value of μ at 835 K, using linear interpolation, is:  3.7995000000000005e-05 N.s/m^2.


### Part 2

Show the system of equations that you must solve to obtain the polynomial coefficients for
a quadratic interpolant suitable for estimating $\mu(835)$. Then solve that system (no need to show
that work - show the solution only) and show the approximation for $\mu(835)$ using a quadratic
polynomial.

The system of equations used to obtain the polynomial coefficients would be the following equations:

$y_1=a_0+a_1\times x_1+a_2\times x_1^2$<br>$y_2=a_0+a_1\times x_2+a_2\times x_2^2$<br>$y_3=a_0+a_1\times x_3+a_2\times x_3^2$

with $x_1=800 K$, $x_2=850 K$, and $x_3=900 K$; and $y_1$, $y_2$, and $y_3$ are the $\mu$ values that correspond to $x_1$, $x_1$, and $x_1$.

In [5]:
import numpy as np
matrix=np.array([[1, 800, 800**2],[1, 850, 850**2],[1, 900, 900**2]])

rhs=np.array([3.698e-05, 3.843e-05, 3.981e-05])

coefficients=np.linalg.solve(matrix, rhs)
print("The coefficients used for quadratic interpolation are a_0=", coefficients[0], ", a_1=", coefficients[1], "and a_2=", coefficients[2],  ", respectively.")

The coefficients used for quadratic interpolation are a_0= 4.25999999999938e-06 , a_1= 5.2100000000001486e-08 and a_2= -1.4000000000000882e-11 , respectively.


Using a quadratic polynomial with the above mentioned polynomial coefficients, the apporoximate value of $\mu(835)=$

In [6]:
y=coefficients[0]*1+coefficients[1]*835+coefficients[2]*835**2
print(y, "N.s/m^2")

3.8002350000000004e-05 N.s/m^2


### Part 3

Create a table summarizing the estimate for $\mu(835)$ using: linear interpolation, quadratic
interpolation, cubic splines, and a polynomial that goes through all of the data points. Be sure to
submit the code that generates these results.

The code used to obtain the approximate value of $\mu(835)$ using linear interpolation and quadratic interpolation is given above. The code used to obtain the approximate value of $\mu(835)$ using cubic splines and a polynomial that goes through all of the data points will be given below.

In [7]:
xi=np.array([ 100, 150, 200, 250, 300, 350, 400, 450, 500, 550, 600, 650, 700, 750, 800,
850, 900, 950, 1000])
yi =np.array([7.110e-06, 1.034e-05, 1.325e-05, 1.596e-05, 1.846e-05, 2.082e-05, 2.301e-05,
2.507e-05, 2.701e-05, 2.884e-05, 3.058e-05, 3.225e-05, 3.388e-05, 3.546e-05,
3.698e-05, 3.843e-05, 3.981e-05, 4.113e-05, 4.244e-5])

from scipy.interpolate import CubicSpline
cs = CubicSpline(xi,yi)
xe=835
ye = cs(xe)
print("The estimated value of μ at 835 K, using cubic spline interpolation, is: ", ye, "N.s/m^2.")


The estimated value of μ at 835 K, using cubic spline interpolation, is:  3.8002442551963886e-05 N.s/m^2.


In [8]:
from numpy import polyfit, polyval
n=len(xi)-1
p = polyfit(xi,yi,n)
xe=835
ye = polyval(p,xe)
print("The estimated value of μ at 835 K, using a polynomial that goes through all of the data points, is: ", ye, "N.s/m^2.")

The estimated value of μ at 835 K, using a polynomial that goes through all of the data points, is:  3.800618026272829e-05 N.s/m^2.


/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:3: RankWarning: Polyfit may be poorly conditioned
  This is separate from the ipykernel package so we can avoid doing imports until


| method                                | approximate value of μ ($value \times10^{-5}$)(N.s/m^2) |
|---------------------------------------|-------------------------------------------|
| linear interpolation                  | 3.7995000000000005                        |
| quadratic interpolation               | 3.8002350000000004                        |
| cubic spline interpolation            | 3.8002442551963886                        |
| polynomial (all points) interpolation | 3.800618026272829                         |

The values in the table indicate that all four methods yield the same answer to within three significant digits, but differ from each other when four or more significant digits are included.

## Problem 2

The atmospheric temperature (in F) as a function of time of day (in hrs), for a typical day in August is
given by:

In [9]:
t=np.array([0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23])
Temperature=np.array([75,74,73,72,71,70,69,69,72,75,79,82,86,88,89,90,91,91,91,90,87,82,80,77])

### Part 1

Plot the Temperature as a function of time.

In [ ]:
%matplotlib inline
import matplotlib.pyplot as plt
plt.xlabel("time (hrs)") # set x label
plt.ylabel("Temperature (degrees F)") # set y label
plt.plot(t,Temperature,"r-",label="Temperature")